### PR TITLE
Feature/ ODP nodes update: calculate and validate dependents nodes

### LINKED ISSUE #1627: ODP nodes update: calculate and validate dependents nodes
when updating / deleting odp,
calculate nodes that are dependents of all variables of forestArea and forestCharacteristics .
it should be the same than updateNodesValues. but with predefined variables.

- [x] on update
- [ ] on delete

### PR DESCRIPTION
**Pull request by minotogna**
part of #1627  and #1666

~~**TODO** Remove originalDataPoint from assessment section data store (feature available in pr #1640)~~

~~**Important** To be merged after #1684~~

https://user-images.githubusercontent.com/1138156/202180505-7aa2104b-19da-4667-a130-0c68020784b0.mov

### CODE REVIEW

File: src/server/controller/cycleData/updateOriginalDataPointDependentNodes.ts [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+import { Assessment, Cycle, OriginalDataPoint } from '@meta/assessment'
+import { NodeUpdates } from '@meta/data'
+import { User } from '@meta/user'
+
+import { BaseProtocol } from '@server/db'
+
+import { calculateAndValidateDependentNodes } from './persistNodeValue/calculateAndValidateDependentNodes'
+
+const dependencies: Array<{ sectionName: string; tableName: string; variableName: string }> = [
+  { sectionName: 'extentOfForest', tableName: 'extentOfForest', variableName: 'forestArea' },
```
Comment (sorja) [inline src/server/controller/cycleData/updateOriginalDataPointDependentNodes.ts]: tableName: TableNames.extentOfForest

Comment (minotogna) [reply src/server/controller/cycleData/updateOriginalDataPointDependentNodes.ts]: sure. you will update this please

File: src/server/controller/cycleData/updateOriginalDataPoint.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import { ActivityLogMessage, Assessment, Cycle, OriginalDataPoint } from '@meta/assessment'
 import { User } from '@meta/user'
 
+import { updateOriginalDataPointDependentNodes } from '@server/controller/cycleData/updateOriginalDataPointDependentNodes'
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint.ts]: relative import here would be preferred

File: src/server/controller/cycleData/updateOriginalDataPoint.ts [hunk 3]
```diff
@@ -16,11 +17,15 @@ export const updateOriginalDataPoint = async (
 ): Promise<OriginalDataPoint> => {
   const { assessment, cycle, originalDataPoint, user } = props
 
+  const updatedOriginalDataPoint = await OriginalDataPointRepository.update(
```
Comment (minotogna) [inline src/server/controller/cycleData/updateOriginalDataPoint.ts]: why is this outside transaction ?

File: src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts [hunk 4]
```diff
@@ -43,17 +45,46 @@ export const validateNodeUpdates = async (props: Props, client: BaseProtocol): P
...
+              { assessment, cycle, countryIso, tableName, variableName, colName, row },
+              client
+            )
+            // eslint-disable-next-line no-await-in-loop
+            const node = await NodeRepository.updateValidation(
+              { assessment, cycle, tableName, variableName, countryIso, colName, validation },
+              client
+            )
+            value = node?.value
+          }
```
Comment (minotogna) [inline src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts]: can u explain what's happening here?

Comment (sorja) [reply src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts]: We fetch the ODP value for the node if the criteria is met:
- table 1a/1b
- year for odp

Comment (minotogna) [reply src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts]: That I could see. But why ?
Can u explain why ?

File: src/server/controller/cycleData/updateOriginalDataPoint.ts (hunk 3 above)
Comment (sorja) [reply src/server/controller/cycleData/updateOriginalDataPoint.ts]: views dont update inside transaction so this needs to be done before, what do you suggest?
(; if we dont move it outside transaction, the updates are 1 state update behind, always showing the previous changed value and calculations)

File: src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts (hunk 4 above)
Comment (sorja) [reply src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts]: - we need to implement odp validation
- normal tables/views work on node table
- odp special case, works on origianl_data_point_data


what do you suggest?
change the repository underneath?
add odps to node table and handle the odps as nodes instead and get rid of specific 'odp' tables?

https://github.com/openforis/fra-platform/blob/97af1228b0cf66f5502b8b4200b6cb834f5fd963/src/server/repository/assessmentCycle/node/updateValidation.ts#L24

Comment (minotogna) [reply src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts]: I think at the moment ODP validation happens only client side, isn't is so?
If yes, here we don't need to validate the ODP.

Comment (sorja) [reply src/server/controller/cycleData/persistNodeValue/validateNodeUpdates/validateNodeUpdates.ts]: so instead, in the parent where validateNodeUpdates is called we instead replace the found values? This one returns the non-odp values anyway, I must look deeper

**Review by sorja — APPROVED** [review]

**Review by rfontanarosa — APPROVED** [review]

**Pull request merged**